libs/remote_storage/Cargo.toml

```diff
@@ -9,8 +9,11 @@ async-trait = "0.1"
 metrics = { version = "0.1", path = "../metrics" }
 utils = { version = "0.1", path = "../utils" }
 once_cell = "1.13.0"
-rusoto_core = "0.48"
-rusoto_s3 = "0.48"
+aws-smithy-http = "0.51.0"
+aws-types = "0.51.0"
+aws-config = { version = "0.51.0", default-features = false, features=["rustls"] }
```

Comment:
Now I'm starting to think: why enable ssl here, and not on `aws-sdk-s3`? Maybe, both are needed?

`rt-tokio` — do you know why that is a default?

Reply:
[Original error](https://github.com/neondatabase/neon/pull/2837) happened here: https://github.com/neondatabase/neon/blob/24d3ed09524b156e6726853aaa14902c695bf039/pageserver/src/storage_sync/download.rs#L148

which looks like the client could also be involved.

Reply:
Yeah it's needed in both but it's already in `aws-sdk-s3` `default-features`, i would day `rt-tokio` it's a default because of the [re-exported ByteStream](https://docs.rs/aws-sdk-s3/0.21.0/aws_sdk_s3/types/struct.ByteStream.html) and it's dependency with tokio, [example](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-smithy-http/src/byte_stream.rs#L400).

Check [this](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-config/src/imds/client.rs#L563), that failed because aws-config was built without the `default-features` and without the `rustls` feature and didn't fail until stage because all the tests use `EnvironmentVariableCredentialsProvider`...

Reply:
I have checked the image that was built by CI for this PR and got another error, still looks like we're using HTTP?

```
2022-11-18T07:28:16.577100Z  WARN storage_sync_loop:storage_sync_loop_step:process_sync_task_batch{sync_id=ffb983546b5e341a2226c12a626ac74f/1ad9b3fd09e5e0119fba1da8920c33e0}:upload_timeline_data: provider failed to provide credentials provider=IAM error=An error occurred while loading credentials: An IO error occurred communicating with IMDS: timeout: error trying to connect: HTTP connect timeout occurred after 1s
2022-11-18T07:28:16.577242Z ERROR storage_sync_loop:storage_sync_loop_step:process_sync_task_batch{sync_id=ffb983546b5e341a2226c12a626ac74f/1ad9b3fd09e5e0119fba1da8920c33e0}:upload_timeline_data: Failed to update remote timeline ffb983546b5e341a2226c12a626ac74f/1ad9b3fd09e5e0119fba1da8920c33e0: Failed to upload new index part
Caused by:
 0: Failed to upload index part for 'ffb983546b5e341a2226c12a626ac74f/1ad9b3fd09e5e0119fba1da8920c33e0'
 1: Failed to upload from '/storage/pageserver/data/tenants/ffb983546b5e341a2226c12a626ac74f/timelines/1ad9b3fd09e5e0119fba1da8920c33e0/index_part.json' to storage path '"zenith-us-stage-ps-4/tenants/ffb983546b5e341a2226c12a626ac74f/timelines/1ad9b3fd09e5e0119fba1da8920c33e0/index_part.json"'
 2: failed to construct request: Failed to load credentials from the credentials provider: An error occurred while loading credentials: An IO error occurred communicating with IMDS: timeout: error trying to connect: HTTP connect timeout occurred after 1s
 3: Failed to load credentials from the credentials provider: An error occurred while loading credentials: An IO error occurred communicating with IMDS: timeout: error trying to connect: HTTP connect timeout occurred after 1s
Stack backtrace:
0: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/result.rs:2105:27
   <remote_storage::s3_bucket::S3Bucket as remote_storage::RemoteStorage>::upload::{{closure}}
          at /home/nonroot/libs/remote_storage/src/s3_bucket.rs:378:9
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
1: <core::pin::Pin<P> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/future.rs:124:9
   remote_storage::GenericRemoteStorage::upload_storage_object::{{closure}}
          at /home/nonroot/libs/remote_storage/src/lib.rs:231:71
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
2: pageserver::storage_sync::upload::upload_index_part::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync/upload.rs:40:10
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
   pageserver::storage_sync::update_remote_data::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:1316:62
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
3: pageserver::storage_sync::upload_timeline_data::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:1195:6
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
   pageserver::storage_sync::process_sync_task_batch::{{closure}}::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:889:26
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
4: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
5: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/future/maybe_done.rs:68:48
6: pageserver::storage_sync::process_sync_task_batch::{{closure}}::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/macros/join.rs:116:24
   <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/future/poll_fn.rs:38:9
7: pageserver::storage_sync::process_sync_task_batch::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:873:44
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
8: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
9: pageserver::storage_sync::process_batches::{{closure}}::{{closure}}::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:832:87
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
10: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/stream/futures_unordered/mod.rs:514:17
   futures_util::stream::stream::StreamExt::poll_next_unpin
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/stream/stream/mod.rs:1626:9
11: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/stream/stream/next.rs:32:9
   pageserver::storage_sync::process_batches::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:841:69
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
12: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
13: pageserver::storage_sync::storage_sync_loop::{{closure}}::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/macros/select.rs:507:49
   <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/future/poll_fn.rs:38:9
14: pageserver::storage_sync::storage_sync_loop::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:723:25
15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
   <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
16: pageserver::storage_sync::spawn_storage_sync_task::{{closure}}
          at /home/nonroot/pageserver/src/storage_sync.rs:687:57
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
17: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:296:9
   <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/future/future/catch_unwind.rs:36:42
   <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
   std::panicking::try::do_call
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
   std::panicking::try
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
   std::panic::catch_unwind
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
   <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/future/future/catch_unwind.rs:36:9
   <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:348:35
   tokio::task::task_local::LocalKey<T>::scope_inner
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:233:19
   <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:345:13
   <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:348:35
   tokio::task::task_local::LocalKey<T>::scope_inner
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:233:19
   <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/task_local.rs:345:13
18: pageserver::task_mgr::task_wrapper::{{closure}}
          at /home/nonroot/pageserver/src/task_mgr.rs:308:10
   <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/future/mod.rs:91:19
19: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:184:17
   tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/loom/std/unsafe_cell.rs:14:9
   tokio::runtime::task::core::CoreStage<T>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/core.rs:174:13
   tokio::runtime::task::harness::poll_future::{{closure}}
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:480:19
   <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
   std::panicking::try::do_call
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
   std::panicking::try
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
   std::panic::catch_unwind
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
   tokio::runtime::task::harness::poll_future
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:468:18
   tokio::runtime::task::harness::Harness<T,S>::poll_inner
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:104:27
   tokio::runtime::task::harness::Harness<T,S>::poll
          at /home/nonroot/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/runtime/task/harness.rs:57:15
20: std::thread::local::LocalKey<T>::with
21: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
22: tokio::runtime::scheduler::multi_thread::worker::Context::run
23: tokio::macros::scoped_tls::ScopedKey<T>::set
24: tokio::runtime::scheduler::multi_thread::worker::run
25: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
26: tokio::runtime::task::harness::Harness<T,S>::poll
27: tokio::runtime::blocking::pool::Inner::run
28: std::sys_common::backtrace::__rust_begin_short_backtrace
29: core::ops::function::FnOnce::call_once{{vtable.shim}}
30: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/alloc/src/boxed.rs:1872:9
   <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/alloc/src/boxed.rs:1872:9
   std::sys::unix::thread::Thread::new::thread_start
          at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/sys/unix/thread.rs:108:17
31: start_thread
          at ./nptl/./nptl/pthread_create.c:477:8
32: __clone
          at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Reply:
That is a fun one, i cannot test the image somehow?
My take on this is the (poorly) defaults chosen, look at that 1s timeout in the exception, i think that comes from [here](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-config/src/imds/client.rs#L51). Rusoto for example uses 30s...
Customising the `imds::Client` requires `async` so i'm looking into options..

Reply:
I still suspect some HTTP quirks rather than timeouts, but timeouts definitely worth a try too.

> i cannot test the image somehow?

You can, every successful PR produces a DockerHub image push: 
https://github.com/neondatabase/neon/actions/runs/3482779648/jobs/5827192299#step:12:21

So, if you use `docker pull neondatabase/neon:3482779648`, that should work: at least, it worked back in the day I've tested it and works for me now, but not sure what are retention policies for such "temporary" images are.

Alas, I don't think I can easily provide you with an infra access, but AFAIK currently that should be very close to  regular "backend behind a AWS IAM" set-up, so you can try it yourself in the AWS infra.

Reply:
If i'm not wrong i would need an EC2 deployed to test the new changes, i didn't test the customising of `imds::Client` and idk if the timeout is the problem i would give the new changes a try...

I don't want to waste your time but i'm blind from my side :(

Reply:
Sorry for not coming back for this, I was a bit ill.

It looks like this time things work, so it was the timeout indeed.
Let's merge the PR? For that, we need two things:
* a rebase
* better way to configure the timeout, another filed in `S3Config` with the default value

Reply:
Oh no problem i hope you are doing well!

First of all thanks for the testing! I submitted an issue in https://github.com/awslabs/aws-sdk-rust/issues/673 because the usage of `provide_credentials_fn` it's not great as a solution of sync scenarios with a customised `imds::Client` because its need for a 'static lifetime, in our case any configuration that gets captured by this `fn` result in a compile error because of the lifetime of `S3Config`.

I reduced the timeout to 10 seconds as `const` because i think it's the simplest solution for now although it's not great.
I expect this as a temporal solution (hopefully) because that second as a timeout i don't think is reasonable.

Reply:
I have tested the build manually again and it works for me still.
I believe, we could merge it again given the reviews in the previous PR.